flink-python/src/main/java/org/apache/flink/python/env/beam/ProcessPythonEnvironmentManager.java

```diff
@@ -163,18 +196,34 @@ public void close() throws Exception {
     }
 
     @Override
-    public PythonEnvironment createEnvironment() throws IOException {
-        Map<String, String> env = constructEnvironmentVariables();
-
-        if (dependencyInfo.getRequirementsFilePath().isPresent()) {
-            LOG.info("Trying to pip install the Python requirements...");
-            PythonEnvironmentManagerUtils.pipInstallRequirements(
-                    dependencyInfo.getRequirementsFilePath().get(),
-                    dependencyInfo.getRequirementsCacheDir().orElse(null),
-                    requirementsDirectory,
-                    dependencyInfo.getPythonExec(),
-                    env);
+    public PythonEnvironment createEnvironment() throws Exception {
+        Map<String, String> env;
+        if (isResponsibleForEnvironment) {
+            env = constructEnvironmentVariables();
+
+            if (dependencyInfo.getRequirementsFilePath().isPresent()) {
+                LOG.info("Trying to pip install the Python requirements...");
+                PythonEnvironmentManagerUtils.pipInstallRequirements(
+                        dependencyInfo.getRequirementsFilePath().get(),
+                        dependencyInfo.getRequirementsCacheDir().orElse(null),
+                        requirementsDirectory,
+                        dependencyInfo.getPythonExec(),
+                        env);
+            }
+            pythonEnvResource.setEnvironmentVariables(env);
+            synchronized (pythonEnvResource) {
+                pythonEnvResource.finish();
```

Comment:
Currently, there are multiple places handling the concurrent access of PythonEnvResource using wait/notify. This is error prone. Could we refactor them a bit to avoid this? SharedResources is a good example on how to achieve this. 

Some thoughts in my mind: All concurrent access of PythonEnvResource should be handled in one place. Maybe we could created class SharedPythonEnvResources and handle all concurrent access in SharedPythonEnvResources.